### DISCUSSION TITLE
can't get grammar to work

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#913] [Q&A] [answered]
Hello! Taking Langium for a spin. I started with the hello-world project template. When I hit F5, it started the extension host. I made a `.hello` file, and that worked great, checking both person and hello statements.

Next I proceeded to modify my `.langium` grammar, but I'm not having luck with the new grammar yet.

Here's the new definition, in case I'm missing something (VS Code shows no errors):

```rb
grammar LanguageExperiment

entry Code:
    (vars+=LetStatement | fns+=Functionn)*;

LetStatement:
    'let' name=ID '=' value=Functionn;

Functionn:
    'function' name=ID '(' ID ')' '{' '}';

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal INT returns number: /[0-9]+/;
terminal STRING: /"[^"]*"|'[^']*'/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

Here's a sample file in the extension host:

```js
let foo = function foo(foo) {};
```

I see these errors:

<img width="366" alt="Screenshot 2023-02-09 at 12 05 25 PM" src="https://user-images.githubusercontent.com/297678/217925723-1f519272-3c15-4ad6-afd3-4259c7d92ce2.png">
<img width="485" alt="Screenshot 2023-02-09 at 12 05 34 PM" src="https://user-images.githubusercontent.com/297678/217925727-1ceb5245-e228-4f9b-bf72-99c763a70b3a.png">
<img width="528" alt="Screenshot 2023-02-09 at 12 07 26 PM" src="https://user-images.githubusercontent.com/297678/217925731-92764de1-940c-4e34-9618-24b57901a411.png">
<img width="548" alt="Screenshot 2023-02-09 at 12 07 35 PM" src="https://user-images.githubusercontent.com/297678/217925734-4c4c49ce-d0b2-4018-87b9-a9bf22e1f22c.png">

In a sibling file I still have

```
person Joe
Hello John!

person John
Hello Joe!
```

and that still works.

Why does the hello example still work, although I've changed the grammar? What's wrong with the new grammar (or the sample "code")?

**Comment by luan-xiaokun** [comment, marked as answer]
Did you run `npm run langium:generate` and `npm run build`? The first command will generate TypeScript files for your DSL, and the second command will compile the TypeScript files to build your extension.
https://github.com/langium/langium/blob/4115491e39e3eb2ad36cbe4223a02ce212d0ecd0/packages/generator-langium/langium-template/package.json#L37-L44

**Reply by trusktr** [reply]
I only hit `f5` like the getting started tutorial, and it opens the separate "extension host" window. I would presume it has everything built right? I'll try some manual steps.

EDIT: I ran `npm run langium:generate` successfully, and then `npm run build` fails with:

```
src/cli/generator.ts:4:10 - error TS2305: Module '"../language-server/generated/ast"' has no exported member 'Model'.

4 import { Model } from '../language-server/generated/ast';
           ~~~~~

src/cli/generator.ts:13:29 - error TS7006: Parameter 'greeting' implicitly has an 'any' type.

13     model.greetings.forEach(greeting => fileNode.append(`console.log('Hello, ${greeting.person.ref?.name}!');`, NL));
                               ~~~~~~~~

src/language-server/lang-validator.ts:2:29 - error TS2305: Module '"./generated/ast"' has no exported member 'Person'.

2 import { LangAstType, Person } from "./generated/ast";
                              ~~~~~~

src/language-server/lang-validator.ts:12:5 - error TS2322: Type '{ Person: (person: Person, accept: ValidationAcceptor) => void; }' is not assignable to type 'ValidationChecks<LangAstType>'.
  Object literal may only specify known properties, and 'Person' does not exist in type 'ValidationChecks<LangAstType>'.

12     Person: validator.checkPersonStartsWithCapital,
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/cli/index.ts:3:10 - error TS2305: Module '"../language-server/generated/ast"' has no exported member 'Model'.

3 import { Model } from "../language-server/generated/ast";
           ~~~~~
```

I'm guessing I need to go an update these files manually when I change my grammar?

**Reply by luan-xiaokun** [reply]
The generated hello-world demo includes a validator ensuring that every `person` after the `Hello` starts with a capital. As you can see, this validator is related to the given hello-world language, so when you change the grammar definition, this function becomes invalid. Therefore, you need to update / remove this function after you begin writing your own grammar. The same is true for `cli/index.ts`. After that, you can play around with the `.langium` grammar definition without updating any TypeScript files.

**Reply by trusktr** [reply]
Nice. All working now! Thanks! Wasn't obvious at first, but makes total sense.

I wonder if we can configure `f5` to also run the type checks, so that we see the errors in the problems tab out of the box. Then as soon as we change the hello-world grammar we'd see the problems right away. 

I am now using `Run Task` with `$tsc-watch` mode, but the tutorial doesn't mention it, so I thought `f5` was gonna provide the whole experience.

**Reply by luan-xiaokun** [reply]
I'm not familiar with VS Code's launch configuration stuff, I just think `f5` as running `./foo` to execute the code, and `npm run build` as `gcc demo.c` to compile the code. I think it's possible to customize your own running procedure by modifying `/.vscode/launch.json`, but I don't know how to do it.

**Reply by msujew** [reply]
You can add a `prelaunchTask` to the extension debug config and call the Langium generator + build script in there. We might add that to the yeoman generator in the future.

**Reply by bnhmn** [reply]
It also took us 2 hours to figure out that you need to run `npm run langium:generate` and `npm run build` in order to apply the changes to the grammar.  
Maybe it makes sense to mention it in the [Getting Started](https://langium.org/docs/getting-started/) page. We looked for it there, but didn't find the information.

After 2-3 days I also figured out that **everything** is explained in the **langium-quickstart.md** file.
If we had seen it earlier it would have saved quite a bit of time. 

Regarding the prelaunchTask thing: I have tested it but it adds some delay to the startup of the Extension Host.
Ideally, there would be a way to automatically run "npm run watch" in the background when opening a Langium project.
But it's also not that hard to run it manually in the terminal.

**Reply by msujew** [reply]
@ben5311 Our getting started page specifically mentions to run `npm run langium:generate` to regenerate the TypeScript files:

![image](https://user-images.githubusercontent.com/4377073/230099477-8fb133a6-d371-43f6-bcda-ca39bf1c6d84.png)

> Regarding the prelaunchTask thing: I have tested it but it adds some delay to the startup of the Extension Host.

Well yes, it's adds security at the cost of time. That's expected.

> Ideally, there would be a way to automatically run "npm run watch" in the background when opening a Langium project.

I'm quite sure that this isn't in the spirit of vscode's security model, given that the `watch` script could contain any kind of code. In fact, it could even be totally unrelated to Langium.

**Reply by bnhmn** [reply]
> Ideally, there would be a way to automatically run "npm run watch" in the background when opening a Langium project.

Actually VS Code can do this. 

Instructions:


1. Make sure concurrently is installed so you can run the two commands as one command: `npm i -D concurrently`.
2. Add / update this npm scripts entry in `package.json`

```json
    "scripts": {
        "watch": "concurrently \"langium generate --watch\" \"tsc -b tsconfig.json --watch\""
    }
```
3. Create the file `.vscode/tasks.json` inside the project folder.

```json
{
    "version": "2.0.0",
    "tasks": [
        {
        "label": "Run npm watch on startup",
        "type": "shell",
        "command": "npm run watch",
        "presentation": {
            "reveal": "always",
            "panel": "new"
        },
        "runOptions": { "runOn": "folderOpen" }
        }
    ]
}
```

Works pretty well. Credits to: https://www.roboleary.net/vscode/2020/10/19/vscode-task-onstartup.html.
